### PR TITLE
refactor: remove TypeORM completely

### PR DESCRIPTION
**Pull request by Nictheboy**

### CODE REVIEW

**Comment by Nictheboy** [comment]
@coderabbitai review

File: src/users/users.prisma [hunk 1]
```diff
@@ -1,23 +1,19 @@
 //
-//  DO NOT MODIFY THIS FILE!
+//  This file is generated by Prisma from the TypeORM schema.
```
Comment (andylizf) [inline src/users/users.prisma:2]: This is obvious enough that it doesn't need to be stated in the document

File: prisma/migrations/20240617031314_refactor_remove_typeorm/migration.sql
Comment (andylizf) [inline prisma/migrations/20240617031314_refactor_remove_typeorm/migration.sql:1]: So you merge `deprecate_answer_attitude` migration and `deprecate_typeorm_on_users` and `refactor_remove_typeorm` into one?

File: src/questions/questions.service.ts [hunk 2]
```diff
@@ -1038,8 +1127,13 @@ export class QuestionsService {
     });
   }
 
-  async getQuestionCount(userId: number): Promise<number> {
-    return await this.questionRepository.countBy({ createdById: userId });
+  getQuestionCount(userId: number): Promise<number> {
+    return this.prismaService.question.count({
+      where: {
+        createdById: userId,
+        deletedAt: null,
```
Comment (andylizf) [inline src/questions/questions.service.ts:1134]: Is there really no need for a middleware here? As mentioned in the [Middleware sample: soft delete
](https://www.prisma.io/docs/orm/prisma-client/client-extensions/middleware/soft-delete-middleware#option-2-use-middleware-to-determine-the-behavior-of-readupdate-queries-for-deleted-records).

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -79,6 +69,16 @@ jobs:
           - 9200:9200
     
     steps:
+      # See: https://remarkablemark.org/blog/2022/05/12/github-actions-postgresql-increase-max-connections-and-shared-buffers/
+      # See: https://stackoverflow.com/questions/70673766/how-to-increase-max-connection-in-github-action-postgres
```
Comment (andylizf) [inline .github/workflows/test.yml:73]: https://stackoverflow.com/a/78546472/8083927 seems much simpler. Does it work?

File: src/answer/answer.service.ts [hunk 4]
```diff
@@ -109,24 +102,31 @@
       );
       return PageHelper.PageStart(currDto, pageSize, (answer) => answer.id);
     } else {
```
[the remark this answers is not in the document]
Comment (andylizf) [reply src/answer/answer.service.ts:104]: Make sense.

**Review by andylizf — CHANGES_REQUESTED** [review]

File: src/answer/answer.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Nictheboy) [reply src/answer/answer.service.ts:104]: This else block appears in paging logic:
```typescript
if (!pageStart) {
    // do something
    return {...}
} else {
    // do something
    return {...}
}
```
Although the else clause can be removed, it is more clear if we leave it hear.

File: prisma/migrations/20240617031314_refactor_remove_typeorm/migration.sql
Comment (Nictheboy) [reply prisma/migrations/20240617031314_refactor_remove_typeorm/migration.sql:1]: It is a mistake. I think `deprecate_answer_attitude` and `deprecate_typeorm_on_users` are both created during this pr, but in fact the first one is not.

File: .github/workflows/test.yml (hunk 3 above)
Comment (Nictheboy) [reply .github/workflows/test.yml:73]: No,it does not work. I've tries it.

File: src/questions/questions.service.ts [hunk 5]
```diff
@@ -94,43 +75,53 @@
...
     });
-    if (relation == null)
+    if (ret.count == 0)
       throw new QuestionNotHasThisTopicError(questionId, topicId);
-    await questionTopicRelationRepository.softRemove({ id: relation.id });
+    /* istanbul ignore if */
+    if (ret.count > 1)
+      throw new Error(
+        `More than one question-topic relation deleted. questionId: ${questionId}, topicId: ${topicId}`,
+      );
```
[the remark this answers is not in the document]
Comment (Nictheboy) [reply src/questions/questions.service.ts]: accepted in commit https://github.com/SageSeekerSociety/cheese-backend/pull/226/commits/d87a714211f1dd9be5f758161c695eb681dea5d4

File: src/groups/groups.service.ts [hunk 6]
```diff
@@ -87,29 +87,47 @@
...
     }
-    if ((await this.groupsRepository.findOneBy({ name })) != undefined) {
+    if (
+      (await this.prismaService.group.count({
+        where: {
+          name,
+          deletedAt: null,
+        },
+      })) > 0
+    ) {
```
[the remark this answers is not in the document]
Comment (Nictheboy) [reply src/groups/groups.service.ts]: accepted in commit https://github.com/SageSeekerSociety/cheese-backend/pull/226/commits/d87a714211f1dd9be5f758161c695eb681dea5d4

File: src/users/users.prisma (hunk 1 above)
Comment (Nictheboy) [reply src/users/users.prisma]: accepted in commit https://github.com/SageSeekerSociety/cheese-backend/pull/226/commits/d87a714211f1dd9be5f758161c695eb681dea5d4

File: prisma/migrations/20240617031314_refactor_remove_typeorm/migration.sql
Comment (Nictheboy) [reply prisma/migrations/20240617031314_refactor_remove_typeorm/migration.sql:1]: accepted in commit https://github.com/SageSeekerSociety/cheese-backend/pull/226/commits/d87a714211f1dd9be5f758161c695eb681dea5d4

File: src/questions/questions.service.ts (hunk 2 above)
Comment (Nictheboy) [reply src/questions/questions.service.ts]: accepted in commit https://github.com/SageSeekerSociety/cheese-backend/pull/226/commits/d87a714211f1dd9be5f758161c695eb681dea5d4

File: src/answer/answer.service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (andylizf) [reply src/answer/answer.service.ts:102]: I think that when the special case can be eliminated early and returned, the rest does not need to be wrapped in an else block. `PageStart==0` is exactly this special case.

File: src/questions/questions.service.ts [hunk 7]
```diff
@@ -94,43 +75,51 @@
     createdById: number,
     // for transaction
     omitQuestionExistsCheck: boolean = false,
-    questionTopicRelationRepository?: Repository<QuestionTopicRelation>,
+    prismaClient?: PrismaClient,
```
[the remark this answers is not in the document]
Comment (andylizf) [reply src/questions/questions.service.ts:78]: The nullish coalescing operator is indeed cleaner and clearer. Additionally, if you decide to adopt the suggestion, you can directly click 'commit suggestion' in the 'Committable Suggestions' collapsible panel.

**Review by andylizf — APPROVED** [review]

**Pull request merged**